src/clients/api/queries/getVTokenApySimulations.ts

```diff
@@ -0,0 +1,86 @@
+import BigNumber from 'bignumber.js';
+
+import { DAYS_PER_YEAR } from 'constants/daysPerYear';
+import { BLOCKS_PER_DAY } from 'constants/blocksPerDay';
+import { COMPOUND_MANTISSA } from 'constants/compoundMantissa';
+import { InterestModel } from 'types/contracts';
+
+const REFERENCE_AMOUNT_WEI = 1e4;
+
+export interface IGetVTokenInterestRatesInput {
+  vTokenInterestModelContract: InterestModel;
+  vTokenReserveFactorMantissa: BigNumber;
+}
+
+export interface VTokenApySnapshot {
+  utilizationRate: number;
+  borrowApyPercentage: number;
+  supplyApyPercentage: number;
+}
+
+export type GetVTokenApySimulationsOutput = VTokenApySnapshot[];
+
+const getVTokenApySimulations = async ({
+  vTokenInterestModelContract,
+  vTokenReserveFactorMantissa,
+}: IGetVTokenInterestRatesInput): Promise<GetVTokenApySimulationsOutput> => {
+  const promises: Promise<VTokenApySnapshot>[] = [];
+
+  for (let u = 1; u <= 100; u++) {
+    const utilizationRate = u / 100;
+
+    const getRates = async () => {
+      const borrowRate = await vTokenInterestModelContract.methods
```

Comment:
Doing that would make this hook a lot more complicated, because if some of these queries fail and we have to retry them then we'll have to track which ones failed and refetch these specifically. It gets even worse when we start thinking about cache invalidation.

The pattern we have in place is to cache every query that needs to be cached. In this case only the resulting data from fetching all the queries should be cached, not the individual queries.

Reply:
I see what you mean, that would solve the cache invalidation issue. I still think it isn't a good idea to separate all these queries into separate queries for the other reason I mentioned however.

Reply:
I don't think we'd achieve what you'd expect, because it would lead to quite a messy code.
The query function would look something like that:
```
const useQueryFn = (uRates: number[]) =>
  useQueries(uRates.map(uRate => ({ queryKey: [QUERY_KEY_BASE, uR], queryFn: fetchUr })))
```

And `useQueryFn` would return an array of query results, which we'd then have to go through to find any potential errors and identify what queries need to be retried. We end up with a lot more complexity than if we treat all the queries as being only one single query.

Reply:
How is that different than troubleshooting the array of promises we have now? With queries we would get the added benefit of caching and graceful error handling, pending. I don't see disadvantages to always calling contracts the same way.

At minimum this should be a hook, We should differentiate between queries which is shaping to be the layer where data is ingested and validated and hooks where data is transformed. This is doing more than just querying and validating data so it should be organized with the hooks.

Reply:
With the array of promises we currently have, if any of the promises throws an error then the whole process is cancelled. The first error is returned and it becomes easy to retry the request.

I'll make the change you're requesting and I think you'll see what I mean. I think it will give us more ground to discuss that.

Reply:
> With the array of promises we currently have, if any of the promises throws an error then the whole process is cancelled. The first error is returned and it becomes easy to retry the request.

That makes sense, I'd think this would be something reconfigurable in `react-query`?

> And about separating the layer that fetches the data from hooks that shape it, this isn't the route we've taken so far. A lot of our query hooks format the data they receive from the contracts before returning. Look at getXvsWithdrawableAmount and getVTokenBalancesAll for exemple.

Yes, I agree we have a clear distinction currently, which is part of the motivation for the comment. To ask about / propose a definition for creating a distinction for those layers